### DISCUSSION TITLE
How to do single execution of script when multiple nodes are selected...

### DISCUSSION THREAD

**Discussion by biswajit-goswami** [freeplane/freeplane#1178] [Help] [answered]
I have a script which opens the below UI for the selected nodes ..its basically to search attributes in the map and assign or delete in the selected nodes:
![image](https://github.com/freeplane/freeplane/assets/11291623/0018bd30-edc5-4e23-8324-141fbab7937e)

The problem is that when multiple nodes are selected and the script is executed, multiple instances of  the UI frame  are created.
eg, if I have 2 nodes selected, then two frames are created...
![image](https://github.com/freeplane/freeplane/assets/11291623/785254b5-b490-41a9-8fe5-3a1e0fbb0d90)

In the script, I am creating a JFrame that renders the UI.
I tried putting all the 3 ExecutionModes ..but same behaviour.

// @ExecutionModes({ON_SELECTED_NODE, ON_SELECTED_NODE_RECURSIVELY})

import javax.swing.*
import java.awt.*
import groovy.swing.SwingBuilder
import javax.swing.table.*
import java.awt.event.*
import java.text.SimpleDateFormat
import java.util.concurrent.TimeUnit

displayUI()

def displayUI(){
swing = new SwingBuilder()
def vars = swing.variables
frame = swing.frame(title: "Search Attributes", id:'searchAttributeFrame',defaultCloseOperation: JFrame.DISPOSE_ON_CLOSE, alwaysOnTop: false,minimumSize: new Dimension(400,25)) {
	panel(){
                     .....
                     ......
         }

        GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
        GraphicsDevice defaultScreen = ge.getDefaultScreenDevice();
        Rectangle rect = defaultScreen.getDefaultConfiguration().getBounds();
        int x = (int) rect.getMaxX() - frame.getWidth()-150;
        int y = 0;
        frame.setLocation(x, y);
        frame.pack()
        frame.show()
}

How do I ensure, only 1 instance of the frame is created?

**Comment by macmarrum** [comment, marked as answer]
You need to run your script for a single node and get your selected nodes using `c.selecteds`, e.g.

```groovy
// @ExecutionModes({ON_SINGLE_NODE})

def selectedNodes = c.selecteds.collect()
```

See also https://docs.freeplane.org/scripting/Scripting.html#execution-modes

**Reply by biswajit-goswami** [reply]
I have tried this execution mode..but still its the same behaviour...

**Reply by macmarrum** [reply]
It's difficult for me to assist without seeing the entire script.
Can you share it here?

**Reply by biswajit-goswami** [reply]
Here it is :

`
// @ExecutionModes({ON_SINGLE_NODE})

import javax.swing.*
import java.awt.*
import groovy.swing.SwingBuilder
import javax.swing.table.*
import java.awt.event.*
import java.text.SimpleDateFormat
import java.util.concurrent.TimeUnit


attributeNames =[]
def selectedNodes = c.selecteds.collect()
//c.select(selectedNodes[0])
initDisplay()
displayUI()

def populateAttributeList(searchText){
        // Search for attribute names matching the search text
        c.findAll().each { node ->
            node.attributes.names.each { attributeName ->
                if (attributeName.contains(searchText)) {
                    if (!attributeNames.contains(attributeName)) {
						attributeNames.add(attributeName)
					}
                }
            }
		}
}
def frame;
def initDisplay(){

swing = new SwingBuilder()
def vars = swing.variables
frame = swing.frame(title: "Search Attributes", id:'searchAttributeFrame',defaultCloseOperation: JFrame.DISPOSE_ON_CLOSE, alwaysOnTop: false,minimumSize: new Dimension(400,25)) {
	panel(){
								boxLayout(axis: BoxLayout.Y_AXIS)
								
								searchAttributePanel=panel(alignmentX: 0f) {
									flowLayout(alignment: FlowLayout.LEFT)
									searcHAttrName=textField(id: 'searchField', columns: 20)
									button(text: "Search", actionPerformed: {
											attributeNames.clear()
											def searchText = vars.searchField.text
											if (searchText && searchText.trim() != "") {
												populateAttributeList(searchText)}
												DefaultListModel model = new DefaultListModel();
												attributeNames.each{
													model.addElement( it );
												}
												recentlist.model = model
												recentspanel.revalidate()
												recentspanel.repaint()
											}
										)
									}
								recentspanel=panel(alignmentX: 0f) {
									flowLayout(alignment: FlowLayout.LEFT)
									//JScrollPane scrollPane = new JScrollPane(JScrollPane.VERTICAL_SCROLLBAR_ALWAYS, JScrollPane.HORIZONTAL_SCROLLBAR_ALWAYS);
									scrollPane {
									//recentlist=list(id: 'type', items: recentNodesVisited)
									recentlist=	list(id: 'type', items: attributeNames, visibleRowCount: 8)
									}
									
									//scrollPane.setViewportView(recentlist);
									MouseListener mouseListener = new MouseAdapter() {
									  public void mouseClicked(MouseEvent mouseEvent) {
										JList theList = (JList) mouseEvent.getSource();
										if (mouseEvent.getClickCount() == 2) {
											selectedNodes.each{selectedNode-> selectedNode.attributes.set(vars.type.selectedValue, "")}
										}
									  }
									};
									recentlist.addMouseListener(mouseListener);

								}
								panel(alignmentX: 0f) {
									flowLayout(alignment: FlowLayout.RIGHT)

								deleteButton = button(action: action(name: 'Delete', defaultButton: true,
								closure: {
								def String attributeToDelete = vars.type.selectedValue
									selectedNodes.each{selectedNode-> selectedNode.attributes.remove(attributeToDelete)}												
								}))
								deleteButton.setMnemonic(KeyEvent.VK_D)
								def closeButton= button(action: action(name: 'Close', closure: {
									//recenthistoryDisplayed=false
									searchAttributeFrame.dispose()
								}))
								closeButton.setMnemonic(KeyEvent.VK_C)
								}
							}
}
        GraphicsEnvironment ge = GraphicsEnvironment.getLocalGraphicsEnvironment();
        GraphicsDevice defaultScreen = ge.getDefaultScreenDevice();
        Rectangle rect = defaultScreen.getDefaultConfiguration().getBounds();
        int x = (int) rect.getMaxX() - frame.getWidth()-150;
        int y = 0;
        frame.setLocation(x, y);
frame.pack()
}
def displayUI(){
if(!frame.isVisible()){
frame.show()
}
}

`

**Reply by macmarrum** [reply]
When I select many nodes and execute the script, it displays a single window.

BTW, you can achieve much the same using `Edit->Node properties->Find and replace attributes…`

**Reply by biswajit-goswami** [reply]
Thanks @macmarrum  for the feedback

So, I restarted FP and then it worked as expected..only 1 frame was created...before the re start, perhaps it did not work as I had tried the other execution modes as well before..and FP would have perhaps retained those settings..not sure though.

Regarding the existing "Find and replace attributes" feature..I am seeking a way to search the attributes using a search box...instead of finding in a drop down..hence was creating a utility for the same.

**Reply by bekoeppel** [reply]
I had the same issue - updated the execution mode of a script, but it only took effect after restarting Freeplane.

The [documentation](https://docs.freeplane.org/scripting/Scripting.html#execution-modes) is actually clear on that:
> Additionally, while changes to script's code are picked up immediately, Freeplane needs to be restarted for @ExecutionModes changes to take effect.
